### PR TITLE
Add H&E images support.

### PR DESCRIPTION
**Pull request by aacic**
## Description

To test this locally the DZI image needs to be generated locally:

`/python/README.md`

An H&E image can be downloaded from here:

https://portal.gdc.cancer.gov/files/73c69d24-6f9e-44e2-bfe5-a608d4cf5c27

`python3 src/dzi.py /path/to/TCGA-22-1017-01Z-00-DX1.9562FE79-A261-42D3-B394-F3E0E2FF7DDA.svs /path/to/tp/files/hg19/pnet/HnEData/dzimage/SJHGG030284_D1/image_name_0`

This command will inside of `/path/to/tp/files/hg19/pnet/HnEData/dzimage/SJHGG030284_D1/` generate :
`image_name_0.dzi `  - dzi config file

`image_name_0` - tiles

In case there is more than one H&E image for the sample, new tiles can be generated also for that image. 

Enable feature toggle in `serverconfig.json`

```
{
  "features": {
    "showDZImages": 1
  }
  ...
}
```

Open the plot:

http://localhost:3000/?noheader=1&mass={%22dslabel%22:%22PNET%22,%22genome%22:%22hg19%22,%22plots%22:[{%22chartType%22:%22matrix%22,%22termgroups%22:[{%22lst%22:[{%22term%22:{%22gene%22:%22TP53%22,%22type%22:%22geneVariant%22}},{%22term%22:{%22gene%22:%22BCOR%22,%22type%22:%22geneVariant%22}},{%22term%22:{%22name%22:%22C19MC%22,%22id%22:%22C19MC%22,%22chr%22:%22chr19%22,%22start%22:54169933,%22stop%22:54292028,%22type%22:%22geneVariant%22}},{%22id%22:%22Gender%22},{%22id%22:%22Age%22,%22q%22:{%22mode%22:%22continuous%22}},{%22id%22:%22TSNE%20Category%22}]}]}]}

Click on the first column in the matrix and click on` H&E images (1)` from the menu.

Number one means there is one H&E image for this sample.

Screenshot 2024-06-04 at 4 45 56 PM

After clicking on it a menu is a sandbox with the openseadragon viewer in sequence mode is opened:

Screenshot 2024-06-05 at 2 37 52 PM

## Checklist

[Check each task](https://github.com/stjude/proteinpaint/wiki/Pull-Request-Checklist) that has been performed or verified to be not applicable.
- [x] Tests: added and/or passed unit and integration tests, or N/A
- [x] Todos: commented or documented, or N/A
- [x] Notable Changes: updated release.txt, prefixed a commit message with "fix:" or "feat:", added to an internal tracking document, or N/A

### CODE REVIEW

File: client/plots/dziviewer/plot.dzi.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+	try {
+		const opts = {
+			holder: holder,
+
+			state: {
+				genome: genomeObj.name,
+				dslabel: dslabel,
+				sample_id: sample_id,
+				dzimage: dzimage,
```
Comment (xzhou82) [inline client/plots/dziviewer/plot.dzi.js:12]: please comment on what is dzimage

File: client/plots/matrix.interactivity.js [hunk 2]
```diff
@@ -256,6 +257,64 @@ export function setInteractivity(self) {
 					self.dom.clickMenu.d.selectAll('*').remove()
 				})
 		}
+
+		const showHnEImages = JSON.parse(sessionStorage.getItem('optionalFeatures') || `{}`)?.showHnEImages
+		if (q.HnEImages && showHnEImages) {
```
Comment (xzhou82) [inline client/plots/matrix.interactivity.js:262]: just a note H&E is one type of such images. there are other belonging to deep zoom category. later the datatype could be renamed to `ds.queries.DZimage{}` with a `imageType='H&E'`, and on client shows q.DZimage.imageType in place of hardcoded `H&E` strings

File: client/plots/DziViewer.js [hunk 3]
```diff
@@ -0,0 +1 @@
+export * from './dziviewer/DziViewer'
```
Comment (xzhou82) [inline client/plots/DziViewer.js:1]: comment on purpose?

File: python/README.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+To run python/src/dzi.py on mac, you need to install the following packages:
+
+brew install python
```
Comment (xzhou82) [inline python/README.md:3]: could you comment on which version you are using?

File: server/routes/sampledzimages.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import fs from 'fs'
+import path from 'path'
+import serverconfig from '#src/serverconfig.js'
+
+/*
+given a sample, return it's singlecell data from dataset
+*/
+
+export const api: any = {
+	endpoint: 'sampledzimages',
```
Comment (xzhou82) [inline server/routes/sampledzimages.ts:10]: what's diff of this to the other `dzimages` route?

**Review by siosonel — DISMISSED** [review]

File: client/plots/dziviewer/plot.dzi.js (hunk 1 above)
Comment (aacic) [reply client/plots/dziviewer/plot.dzi.js]: Added comments.

File: server/routes/dzimages.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+		let imagePath
+		try {
+			const g = genomes[req.query.genome]
+			if (!g) throw 'invalid genome name'
+			const ds = g.datasets[req.query.dslabel]
+			if (!ds) throw 'invalid dataset name'
+
+			imagePath = path.join(
+				`${serverconfig.tpmasterdir}/${ds.queries.HnEImages.samples}`,
+				req.path.split(`/${routePath}`)[1]
```
[the remark this answers is not in the document]
Comment (aacic) [reply server/routes/dzimages.ts]: @siosonel fixed this and `sampleid` is now passed as the route path. The path is sanitized on the backed side too.

File: client/plots/DziViewer.js (hunk 3 above)
Comment (aacic) [reply client/plots/DziViewer.js:2]: I've added a comment.

File: server/routes/sampledzimages.ts (hunk 5 above)
Comment (aacic) [reply server/routes/sampledzimages.ts:10]: This route loads all dzi images paths for specified sample_id and dataset. The other route loads the tiles based on the path provided by this endpoint.

File: client/plots/matrix.interactivity.js [hunk 7]
```diff
@@ -256,6 +257,48 @@ export function setInteractivity(self) {
 					self.dom.clickMenu.d.selectAll('*').remove()
 				})
 		}
+
+		const showDZImages = JSON.parse(sessionStorage.getItem('optionalFeatures') || `{}`)?.showDZImages
+		if (q.DZImages && showDZImages) {
+			async function fetchHnEImages() {
+				return dofetch3('sampledzimages', {
```
Comment (xzhou82) [inline client/plots/matrix.interactivity.js:264]: why not just `const images= await dofetch...`

File: server/routes/sampledzimages.ts (hunk 5 above)
Comment (xzhou82) [reply server/routes/sampledzimages.ts:10]: then please add this comment to top and replace the borrowed one

File: server/routes/termdb.config.ts [hunk 8]
```diff
@@ -206,13 +206,21 @@ function addGenomicQueries(c, ds, genome) {
 		}
 	}
 
+	if (q.DZImages) {
+		q2.DZImages = {}
+		for (const k in q.DZImages) {
+			q2.DZImages[k] = JSON.parse(JSON.stringify(q.DZImages[k]))
```
Comment (xzhou82) [inline server/routes/termdb.config.ts:212]: should only reveal `type` property but not folder path

File: client/plots/matrix.interactivity.js (hunk 7 above)
Comment (siosonel) [reply client/plots/matrix.interactivity.js:264]: @aacic the `async` keyword should be used in the function declaration above, `self.mouseclick = function (event, data) {`, and not be declaring an async function inside this function, so you can then simply do `const images = awaif dofetch3 (...)`

File: server/routes/sampledzimages.ts (hunk 5 above)
Comment (aacic) [reply server/routes/sampledzimages.ts:10]: Changed the comment.

File: client/plots/matrix.interactivity.js (hunk 7 above)
Comment (aacic) [reply client/plots/matrix.interactivity.js]: Refactored.

File: server/routes/termdb.config.ts (hunk 8 above)
Comment (aacic) [reply server/routes/termdb.config.ts]: Fixed.

**Review by xzhou82 — APPROVED** [review]
thanks for prototyping

**Pull request merged**